devito/propagator.py

```diff
@@ -178,8 +212,17 @@ def generate_space_loops_blocking(self, loop_body):
         ivdep = True
         remainder = False
         orig_loop_body = loop_body
+
+        inner_most_dim_passed = False  # used in oder to avoid at inner most dimension
+
         for spc_var, block_size in reversed(zip(list(self.space_dims), self.block_sizes)):
             dim_var = str(self._var_map[spc_var])
+
+            if self.auto_tune and inner_most_dim_passed:  # change block size into var
```

Comment:
When you convert block size into var, you are breaking the assumption in line 242 below. That line assumes the block size is known at "python run time" and goes into the C code as a constant. That block size is used to calculate the size of the remainder loop. In the current code you use the first-guess block size (the "optimal" block size) for that calculation, while you are actually varying it later. This leads to **incorrect number of loop iterations**.